lib/StreamingStore.ts

```diff
@@ -40,14 +40,65 @@ implements RDF.Source<Q>, RDF.Sink<RDF.Stream<Q>, EventEmitter> {
     }
   }
 
-  protected importToListeners(stream: RDF.Stream<Q>): void {
-    stream.on('data', (quad: Q) => {
-      for (const pendingStream of this.pendingStreams.getPendingStreamsForQuad(quad)) {
-        if (!this.ended) {
-          pendingStream.push(quad);
-          pendingStream.emit('quad', quad);
-        }
+  /**
+   * This function will read the quad stream in an on-demand fashion, and will check if the quads already exist in the
+   * store. If they don't, they will be pushed into the storeImportStream, and the matching pendingStreams.
+   * @param stream A quad stream.
+   * @param storeImportStream A stream to import the quads into the store.
+   */
+  protected importToListeners(stream: RDF.Stream<Q>, storeImportStream: PassThrough): void {
+    let streamEnded = false;
+    stream.once('readable', async() => {
+      let quad: Q | null = stream.read();
+      while (quad) {
+        const staticQuad = quad;
+        await new Promise<void>(resolve => {
+          // Match the new quad with the store.
+          const matchStream = this.store.match(
+            staticQuad.subject,
+            staticQuad.predicate,
+            staticQuad.object,
+            staticQuad.graph,
+          );
+
+          // If the StreamingStore hasn't ended, we add the quad to the storeImportStream and the corresponding
+          // pendingStreams and resolve to handle the next quad.
+          const handleEnd = (): void => {
+            if (!this.ended) {
+              storeImportStream.push(staticQuad);
+              for (const pendingStream of this.pendingStreams.getPendingStreamsForQuad(staticQuad)) {
+                pendingStream.push(staticQuad);
+                pendingStream.emit('quad', staticQuad);
+              }
+            }
+            resolve();
+          };
+
+          // If the matchStream has a result, the quad already exists.
+          // We remove the 'end' listener and continue to the next quad.
+          matchStream.once('data', () => {
+            matchStream.removeListener('end', handleEnd);
+            resolve();
+          });
+
+          // If the matchStream has ended (and this listener isn't removed), the quad doesn't exist yet.
+          // So we call the handleEnd function.
+          matchStream.once('end', handleEnd);
+        });
+
+        quad = stream.read();
       }
+      // If the stream has ended, all quads will be read from the quad stream, so we can end the storeImportStream.
+      if (streamEnded) {
+        storeImportStream.end();
+        return;
+      }
+      // If the stream hasn't ended, we recursively call this function to wait for the stream to become readable again.
```

Comment:
After reading this comment, I'm wondering if this recursion is still needed if we were to use `on` instead of `once` on line 51?

Related to this, we may also want to add a unit test to check for the case where the readable-event is invoked multiple times. The test on line 482 goes a bit in this direction, but I'm not sure if readable will actually be invoked multiple times there. We may have to resort to timeouts (which are simulated through jest, so we can manually step through them).

Reply:
@maartyman Could you have a look to see if this is needed?
Would be good to get this one merged :-)

Reply:
If `on` is used, test `handles duplicates in import (set-semantics)` fails. This is due to the next match being called on the store before the previous Quad is added. If I can fix it by awaiting a timeout of 0 ms after adding the quad on line 68.

Reply:
^ nvm this is very mutch not reliable, sometimes the timing works out, sometimes it doesn't.

Reply:
> Related to this, we may also want to add a unit test to check for the case where the readable-event is invoked multiple times. The test on line 482 goes a bit in this direction, but I'm not sure if readable will actually be invoked multiple times there. We may have to resort to timeouts (which are simulated through jest, so we can manually step through them).

It does invoke readable multiple times.

Reply:
Ok, thanks for checking, so no changes are needed then I guess?

Reply:
No, everything should work!